### PR TITLE
Fix double precision + ddp_spawn

### LINKED ISSUE #6893: DDP_SPAWN + Double Precision gives pickle error
## 🐛 Bug

Pickle error with double precision + DDP-spawn

`Can't pickle <function BoringModel.training_step at 0x7f8d684573b0>: it's not the same object as __main__.BoringModel.training_step`

Fixing general pickle error may also solve discussion #6851

### To Reproduce

Use `bug_report_model.py`, add `accelerator='ddp-spawn'` and `precision=64`

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #6893 - should also prevent issue from discussion #6851

Basically just changes patching to wrapping, had to add a new base wrapper class and support for unwrapping, but it means that double precision + distributed wrappers works fine and no longer needs a teardown.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/overrides/base.py [hunk 1]
```diff
@@ -19,25 +21,61 @@
 from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
 
 
-class _LightningModuleWrapperBase(DeviceDtypeModuleMixin, torch.nn.Module):
+class _LightningPrecisionModuleWrapperBase(torch.nn.Module):
```
Comment (justusschock) [inline pytorch_lightning/overrides/base.py:24]: Do you also want to do a fully fp16 based version (instead of only the AMP thing pytorch and apex provide) or why did you also introduce a base class for this? If not, we probably don't need the base class as of now.

Comment (justusschock) [inline pytorch_lightning/overrides/base.py:24]: Also maybe we should also derive this from `DeviceDtypeMixin` since I am not sure, all the device and dtype changes are forwarded correctly (I think per default pytorch only traverses tensors (including children tensors) but not the modules themselves, meaning that the properties would not be updated. Mind also adding a test for this?

**Review by justusschock — COMMENTED** [review]
I like it, some small comments :)

File: pytorch_lightning/overrides/base.py (hunk 1 above)
Comment (ethanwharris) [reply pytorch_lightning/overrides/base.py]: Good catch! I've added some tests :smiley:

Comment (ethanwharris) [reply pytorch_lightning/overrides/base.py]: The reason to have a base wrapper is that the `unwrap_lightning_module` method needs to know how to unwrap both wrapper types. The other option would be to just make the double precision wrapper extend `_LightningModuleWrapperBase` and make the unwrap method recursive

**Comment by ethanwharris** [comment]
@justusschock Thanks for the review, I finally got round to addressing your comments :smiley:

File: pytorch_lightning/overrides/base.py [hunk 2]
```diff
@@ -19,28 +21,70 @@
 from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
 
 
-class _LightningModuleWrapperBase(DeviceDtypeModuleMixin, torch.nn.Module):
+class _LightningPrecisionModuleWrapperBase(DeviceDtypeModuleMixin, torch.nn.Module):
```
Comment (awaelchli) [inline pytorch_lightning/overrides/base.py:24]: so a total noob question but why do we need this wrapping just for 64 bit? Why didn't 16 bit have it until now? Do we need it even if amp support 64 bit?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 3]
```diff
@@ -150,19 +150,19 @@ def start_predicting(self, trainer: 'pl.Trainer') -> None:
...
-        return self.lightning_module.validation_step(*args, **kwargs)
+        return self.model.validation_step(*args, **kwargs)
 
     def test_step(self, *args, **kwargs):
-        return self.lightning_module.test_step(*args, **kwargs)
+        return self.model.test_step(*args, **kwargs)
 
     def predict_step(self, *args, **kwargs):
-        return self.lightning_module.predict_step(*args, **kwargs)
+        return self.model.predict_step(*args, **kwargs)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:165]: Can you give it a quick look again double check that all plugins that don't override these methods actually refer to the right model/wrapper here and it aligns with your intentions?
So in DDP fore example, self.model would be the DistributedDataParallel.

File: pytorch_lightning/plugins/precision/double.py [hunk 4]
```diff
@@ -43,55 +44,63 @@ def _to_double_precision(data: torch.Tensor) -> torch.Tensor:
...
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        return self.module.forward(
+            *LightningDoublePrecisionModule._move_float_tensors_to_double(args),
+            **LightningDoublePrecisionModule._move_float_tensors_to_double(kwargs),
+        )
 
 
 class DoublePrecisionPlugin(PrecisionPlugin):
-    """Plugin for training with double (``torch.float64``) precision."""
+    """ Plugin for training with double (``torch.float64``) precision. """
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/double.py:85]: I like this change!

**Review by awaelchli — COMMENTED** [review]
the test are nice!

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 3 above)
Comment (ethanwharris) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py]: Yeah, can't remember the reason for this and I think it's an error, reverted!

File: pytorch_lightning/overrides/base.py (hunk 2 above)
Comment (ethanwharris) [reply pytorch_lightning/overrides/base.py:24]: The other precision things like AMP just do their own patching so don't need the wrapper. But they don't support 64 bit as they are mainly about trying to avoid floating point errors wioth low precision, not a problem with 64 bit. We only need the wrapping for 64 bit so that we can cast things correctly but still be pickleable for DDP spawn.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 3 above)
Comment (ethanwharris) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:165]: Ok, I've done some digging. All training type plugins either don't wrap the model (so `self.model == self.lightning_module`) or override the methods to call `self.model`. So I've removed the overrides now and put it back to using `self.model` in the base class :smiley:

**Comment by awaelchli** [comment]
Do we wrap the model at the right place? I'm looking at the DDPPlugin, and in the configure_ddp method we have:

`DistributedDataParallel(LightningDistributedModule(LightningModule))))`

Where do you want your precision wrapper to be inserted?
LightningDistributedModule maps forward to the step methods, so I guess you want this:

Looking at the code, you expect this:
`DistributedDataParallel(PrecisionWrapper(LightningDistributedModule(LightningModule))))`

But I don't think that's what's happening. I could be wrong.

**Comment by ethanwharris** [comment]
@awaelchli The precision plugin is set-up here in Accelerator: https://github.com/PyTorchLightning/pytorch-lightning/blob/ac92b57e2b10967d778bae65ecad138df5926f6d/pytorch_lightning/accelerators/accelerator.py#L92

So the wrapping will be PrecisionPlugin(DDPPlugin(...)), I had thought they were the other way around but this still works with the unwrap function

**Comment by ethanwharris** [comment]
@awaelchli I've figured it out, I was right the first time haha. DDP configure actually gets called in the `pre_dispatch` hook, whereas prepcision plugin is called in setup. So the correct wrapping order is actually DDP(DDPPlugin(PrecisionPlugin)))

**Review by Borda — DISMISSED** [review]
can you reconsider this PR with respect to concurrent #7289?

**Comment by ethanwharris** [comment]
@Borda could explain how #7289 is concurrent, that's to do with wrapping `DataModule` right? The fix here is to do with wrapping `Plugin`

**Comment by Borda** [comment]
> @Borda could explain how #7289 is concurrent, that's to do with wrapping `DataModule` right? The fix here is to do with wrapping `LightningModule` for plugins

yep, but shall we also rather use the `__new__` construct?

**Comment by ethanwharris** [comment]
@Borda I don't think we can use `__new__` here is we just want to sometimes wrap the module. `__new__` would work if we wanted to always wrap it (e.g. to decorate some of the functions).

**Comment by ethanwharris** [comment]
@Borda @awaelchli @justusschock I've updated the PR, mind re-reviewing?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/precision/double.py [hunk 5]
```diff
@@ -43,55 +44,63 @@ def _to_double_precision(data: torch.Tensor) -> torch.Tensor:
...
+        )
+
+    def predict_step(self, *args: Any, **kwargs: Any) -> Any:
+        return self.module.predict_step(
+            *LightningDoublePrecisionModule._move_float_tensors_to_double(args),
+            **LightningDoublePrecisionModule._move_float_tensors_to_double(kwargs),
+        )
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        return self.module.forward(
```
Comment (justusschock) [inline pytorch_lightning/plugins/precision/double.py:78]: maybe call `self.module` instead of `self.module.forward`? In general I'd say this would be fine to directly call forward from submodules, but here I'd prefer to call the module itself.

This is because the module's `__call__` calls forward + all attached hooks. However, when attaching hooks, the user would most likely do this on the `LightningModule` he defined and not on the wrapper we dynamically attach. So when we only call `forward` here, we might lose them.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/precision/double.py (hunk 5 above)
Comment (ethanwharris) [reply pytorch_lightning/plugins/precision/double.py]: Done :smiley:

File: tests/plugins/test_double_plugin.py [hunk 6]
```diff
@@ -134,4 +138,29 @@ def test_double_precision(tmpdir, boring_model):
     trainer.test(model)
     trainer.predict(model)
 
-    assert model.training_step == original_training_step
+
+@RunIf(min_gpus=2)
+@pytest.mark.parametrize('ddp_backend', ('ddp', 'ddp_spawn'))
```
Comment (awaelchli) [inline tests/plugins/test_double_plugin.py:143]: won't work. 

a) remove ddp
b) add special=True to RunIf

considering that this pr aims at fixing a pickle test, I recommend going for a) or even just merge the test with the above.

Comment (ethanwharris) [reply tests/plugins/test_double_plugin.py]: Done :smiley:

**Review by awaelchli — APPROVED** [review]
Note the test for double precision now takes 70+ seconds because of spawn getting called six times 2 x (fit, test, predict)
This is not ideal especially for a test that does not assert anything.

**Comment by ethanwharris** [comment]
@awaelchli I've simplified the test so now should only spawn once :smiley:
